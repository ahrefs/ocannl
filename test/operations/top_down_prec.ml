(* A simple demo of using the Threefry4x32 PRNG in OCANNL *)

open Base
module Tensor = Ocannl.Tensor
module Train = Ocannl.Train
module TDSL = Ocannl.Operation.TDSL
module Tn = Ir.Tnode

let () =
  Tensor.unsafe_reinitialize ();
  let module Backend = (val Backends.fresh_backend ()) in
  let%op d = ("a" [ 2 ] + "b" [ 2 ]) *. "c" [ 2 ] in
  Tn.update_prec b.value Ir.Ops.half;
  Tn.update_prec d.value Ir.Ops.bfloat16;
  (* Even when the default precision is single, c is bfloat16 and a is half. *)
  Ocannl.Train.set_hosted d.value;
  ignore (Ocannl.Train.forward_once (module Backend) d);
  Train.printf_tree d
