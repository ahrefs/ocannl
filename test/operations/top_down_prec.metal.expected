#include <metal_stdlib>
using namespace metal;
kernel void d_fwd(
    device bfloat *c [[buffer(0)]],
    device bfloat *d [[buffer(1)]],
    device half *a [[buffer(2)]],
    device half *b [[buffer(3)]],
    uint3 gid [[threadgroup_position_in_grid]],
    uint3 lid [[thread_position_in_threadgroup]]) {

  /* Local declarations and initialization. */

  /* Main logic. */
  /* d fwd */
  d[0] = (((bfloat)(a[0]) + (bfloat)(b[0])) * c[0]);
  /* end */
}